datameta/security/authz.py

```diff
@@ -0,0 +1,110 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+def user_is_target(user, target_user):
+    return user.id == target_user.id
+
+def has_group_rights(user, group):
+    return user.site_admin or (user.group_admin and user.group.id == group.id)
+
+def is_power_grab(user, target_user):
+    return not user.site_admin and target_user.site_admin
+
+def has_data_access(user, data_user_id, data_group_id=None, was_submitted=False):
+    # if dataset was already submitted, the group must match, or the user must have site_read priviledges
+    # if dataset was not yet submitted, the user must match
+    return any((
+        (user.site_read),
+        (was_submitted and data_group_id and data_group_id == user.group_id),
+        (not was_submitted and data_user_id and data_user_id == user.id)
+    ))
+
+
+def view_apikey(user, target_user):
+    return user_is_target(user, target_user)
+
+def delete_apikey(user, target_user):
+    return user_is_target(user, target_user)
+
+def view_appsettings(user):
+    return user.site_admin
+
+def update_appsettings(user):
+    return user.site_admin
+
+def submit_file(user, file_obj):
+    return has_data_access(user, file_obj.user_id)
+
+def view_file(user, file_obj):
+    was_submitted = (
+        file_obj.content_uploaded and 
+        file_obj.metadatumrecord and  
+        file_obj.metadatumrecord.metadataset.submission_id 
+    )
+    group_id = file_obj.metadatumrecord.metadataset.submission.group_id if was_submitted else None
+    return has_data_access(user, file_obj.user_id, group_id, was_submitted=was_submitted)
+
+def change_groupname(user):
+    return user.site_admin
+
+def view_group_submissions(user, group_id):
+    return any((
+        group_id in [user.group.uuid, user.group.site_id],
+        user.site_read
+    ))
+
+def update_metadatum(user):
+    return user.site_admin
+
+def create_metadatum(user):
+    return user.site_admin
+
+def submit_mset(user, mds_obj):
+    return has_data_access(user, mds_obj.user_id)
+
+def delete_mset(user, mdata_set):
+    return user.id == mdata_set.user_id
+    
+def view_mset(user, mds_obj):
+    was_submitted = bool(mds_obj.submission_id is not None)
+    group_id = mds_obj.submission.group_id if was_submitted else None
+    return has_data_access(user, mds_obj.user_id, data_group_id=group_id, was_submitted=was_submitted)
+    
+def change_password(user, target_user):
+    return user_is_target(user, target_user)
+
+def change_group(user):
+    return user.site_admin
```

Comment:
As mentioned in the previous PR, it would be good if we used a more systematic naming schema which - in case the check is for a specific field - mentions also the entity / holder. Also, we're using two verbs (`change` and `update`) for the same action.
```
change_group
```
could be come
```
update_user_group
```
for example. `change_group` does not specify the entity, it could also mean to change the group of a submission, metadataset or other entities that have a `group` attribute.